### ISSUE TITLE
DllNotFoundException: GfxPluginNativeRender-x64 assembly:<unknown assembly> type:<unknown type> member:(null)

### ISSUE THREAD

**Issue by 314bytes** [PlayEveryWare/eos_plugin_for_unity#465]
**Describe the bug**
Through Unity Analytics, I observed that a couple of players of my videogame have encountered the following error:
```
DllNotFoundException: GfxPluginNativeRender-x64 assembly:<unknown assembly> type:<unknown type> member:(null)
PlayEveryWare.EpicOnlineServices.EOSManager+EOSSingleton.GetEOSPlatformInterface () (at <912b311dc55c467ca19c5df8e28ac756>:0)
PlayEveryWare.EpicOnlineServices.EOSManager+EOSSingleton.OnShutdown () (at <912b311dc55c467ca19c5df8e28ac756>:0)
PlayEveryWare.EpicOnlineServices.EOSManager.OnApplicationQuit () (at <912b311dc55c467ca19c5df8e28ac756>:0)
```

This issue appears to occur during shutdown, but neither I nor my QA team has been able to reproduce it.
I thought to report it and check with you if you can think of any reasons why some users might be experiencing this.

I've noticed that this crash seems to randomly happen with users who have older graphics cards. For instance, the specifications of the last occurrence are as follows:

- Windows 10 (10.0.19045) 64bit
- AMD Ryzen 5 3600 6-Core Processor
- NVIDIA GeForce GT 730

However, I'm puzzled about why an old graphics card would cause a `DllNotFoundException`. Also, since it throws during shutdown, it implies that they were able to load the game correctly, and `GetEOSPlatformInterface()` had worked in all previous calls.

Anyway, I've been wondering if there is something that you can change in EOSManager.cs to handle this situation more gracefully. Perhaps incorporating a try/catch block might be a solution?

**To Reproduce**
I have no idea...

**Expected behavior**
PlayEverywhere's EOS plugin doesn't throw exceptions during shutdown.

**Desktop:**
 - OS: Windows 10
 - Standalone build
 - Unity 2021.3.30f1
 - Plugin 2.3.3

**Comment by paulhazen** [comment]
I @314bytes. My first go-to suggestion would be to update the version of the plugin you are using, since we have made major releases since then, and it's very likely that the issue was resolved in a release subsequent to the one you have in production. It will otherwise be difficult to diagnose unfortunately.

Let me know if you need any direction on upgrading your version of the plugin, we have docs for that :D

**Comment by arthur740212** [comment]
In addition to updating the plug-in.
One clue we noticed is that a couple of players were facing the issue, which leads us to suspect that the visual studio runtime redistributables weren’t installed on their devices.

Check out this related issue for more information.
https://github.com/PlayEveryWare/eos_plugin_for_unity/issues/329

**Comment by 314bytes** [comment]
Thank you both very much! I'll give the suggestions a try.
Currently, I'm not installing any distributable with the game, so that might be the reason....

**Comment by 314bytes** [comment]
@paulhazen I have upgraded the plugin to the latest version and will deploy the upgrade to the users in the next 1-2 days, as soon as I have done some testing.
However I've diffed the source code of the native DLL between the version 2.3.3 and the latest 3.0.2 and it is identical except for the addition of a boolean server flag, which doesn't have any impact for this issue.

I don't think it is an issue with the redistributables. If a DLL doesn't find the right redistributable, it crashes immediately as soon as it is called. In my case however the DLL crashes randomly while the player is already in the game since several minutes, or even at shutdown.
The function `GetEOSPlatformInterface()`, the function that is crashing, is called as soon as the game awakes.

For instance, this is the last exception, just recorded few hours ago:
```
PlayEveryWare.EpicOnlineServices.EOSManager+EOSSingleton.GetEOSPlatformInterface () (at <912b311dc55c467ca19c5df8e28ac756>:0)
PlayEveryWare.EpicOnlineServices.Samples.EOSAchievementManager.GetEOSAchievementInterface () (at <4d21c344bfa142ee947623374bc80a95>:0)
PlayEveryWare.EpicOnlineServices.Samples.EOSAchievementManager.UnlockAchievementManually (System.String achievementId, Epic.OnlineServices.Achievements.OnUnlockAchievementsCompleteCallback callback) (at <4d21c344bfa142ee947623374bc80a95>:0)
EOSSystem.OnMMEvent (MM.Tools.MMAchievementUnlockedEvent achievementUnlockedEvent) (at <4d21c344bfa142ee947623374bc80a95>:0)
MM.Tools.MMEventManager.TriggerEvent[MMEvent] (MMEvent newEvent) (at <4383925b68224e92ac5510c9e375bf27>:0)
MM.Tools.MMAchievementUnlockedEvent.Trigger (MM.Tools.MMAchievement newAchievement) (at <4383925b68224e92ac5510c9e375bf27>:0)
MM.Tools.MMAchievement.UnlockAchievement () (at <4383925b68224e92ac5510c9e375bf27>:0)
MM.Tools.MMAchievement.EvaluateProgress () (at <4383925b68224e92ac5510c9e375bf27>:0)
MM.Tools.MMAchievement.AddProgress (System.Int32 newProgress) (at <4383925b68224e92ac5510c9e375bf27>:0)
MM.Tools.MMAchievementManager.AddProgress (System.String achievementID, System.Int32 newProgress) (at <4383925b68224e92ac5510c9e375bf27>:0)
CustomAchievements.OnMMEvent (CollectibleEvent e) (at <4d21c344bfa142ee947623374bc80a95>:0)
MM.Tools.MMEventManager.TriggerEvent[MMEvent] (MMEvent newEvent) (at <4383925b68224e92ac5510c9e375bf27>:0)
CollectibleEvent.Trigger (System.String name) (at <4d21c344bfa142ee947623374bc80a95>:0)
CustomCollectibleItem.Interact () (at <4d21c344bfa142ee947623374bc80a95>:0)
SC.Character.TryInteract () (at <4d21c344bfa142ee947623374bc80a95>:0)
SC.Character.Update () (at <4d21c344bfa142ee947623374bc80a95>:0)
```

From the callstack it is clear that the player was interacting with an item in the game which triggered an achievement unlock action but, as soon as the function `GetEOSPlatformInterface()` was called, they had an exception. _I guess they weren't too happy ahaha_

By the time the player is able to interact with anything in the game, the Native Gfx plugin has been already initialized since a looong time, so I think there might be something crashing inside the DLL.
Is there anything I can do from my side to help determine what is wrong?

**Comment by 314bytes** [comment]
@arthur740212 you were right, it was a redistributable issue! 

I thought it was impossible because I was getting exceptions reported while the user was already playing or even at shutdown (while the Plugin Gfx DLL is initialized at the game startup) but, after testing, I realized that this kind of exceptions do not block the game execution, so now I'm convinced that the same player simply got more than 1 exception and I'm just seeing the log of the last one.

Anyways, I managed to reproduce exactly the same exception of 1 player by running the game in a Windows Sandbox, in which there were no redistributable installed by default. Then after I installed the latest redist and run again the game, the exception was gone.

Thank you so much!
